app/models/operatingsystem.rb

```diff
@@ -260,12 +263,17 @@ def deduce_family
     end
   end
 
-  def boot_files_uri(medium, architecture, host = nil)
-    raise ::Foreman::Exception.new(N_("%{os} medium was not set for host '%{host}'"), :host => host, :os => self) if medium.nil?
-    raise ::Foreman::Exception.new(N_("Invalid medium '%{medium}' for '%{os}'"), :medium => medium, :os => self) unless media.include?(medium)
-    raise ::Foreman::Exception.new(N_("Invalid architecture '%{arch}' for '%{os}'"), :arch => architecture, :os => self) unless architectures.include?(architecture)
-    self.family.constantize::PXEFILES.values.map do |img|
-      medium_vars_to_uri("#{medium.path}/#{pxedir}/#{img}", architecture.name, self)
+  def boot_files_uri(medium_provider, &block)
+    boot_file_sources(medium_provider, &block).values
+  end
```

Comment:
So in discovery during orchestration I need to find URLs for kexec, what I do:

```ruby
@kexec_kernel, @kexec_initrd = operatingsystem.boot_files_uri(@host.medium, @host.architecture, @host)
```

With this patch I need to do:

```ruby
provider = Foreman::Plugin.medium_providers.find_provider(@host)
@kexec_kernel, @kexec_initrd = operatingsystem.boot_files_uri(provider)
```

Reply:
You will be able to use shorter version:

``` ruby
@kexec_kernel, @kexec_initrd = operatingsystem.boot_files_uri(@host.host_medium_provider)
```